src/reducers/computeRenderedColumns.js

```diff
@@ -0,0 +1,244 @@
+ * positions in the column buffer.
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   endBufferIdy: number,
+ *   endViewportIdy: number,
+ *   firstBufferIdy: number,
+ *   firstViewportIdy: number,
+ * }} columnRange
+ * @param {boolean} viewportOnly
+ * @private
+ */
+function computeRenderedColumnOffsets(state, columnRange, viewportOnly) {
+  const { columnBufferSet, columnOffsetIntervalTree } = state;
+  const {
+    endBufferIdy,
+    endViewportIdy,
+    firstBufferIdy,
+    firstViewportIdy,
+  } = columnRange;
+
+  const {
+    fixedColumnOffsets,
+    fixedRightColumnOffsets,
+    columnGroupOffsets,
+    fixedColumnGroupOffsets,
+    fixedRightColumnGroupOffsets,
+  } = columnWidths(state);
+
+  // we are only doing calculations for scrollable columns
+  state.columnGroupOffsets = columnGroupOffsets;
+  state.fixedColumnOffsets = fixedColumnOffsets;
+  state.fixedRightColumnOffsets = fixedRightColumnOffsets;
+  state.fixedColumnGroupOffsets = fixedColumnGroupOffsets;
+  state.fixedRightColumnGroupOffsets = fixedRightColumnGroupOffsets;
+
+  const renderedColumnsCount = endBufferIdy - firstBufferIdy;
+  if (renderedColumnsCount === 0) {
+    state.columnOffsets = {};
+    state.columnsToRender = [];
+    return;
+  }
+
+  const startIdx = viewportOnly ? firstViewportIdy : firstBufferIdy;
+  const endIdx = viewportOnly ? endViewportIdy : endBufferIdy;
+
+  // output for this function
+  const columns = []; // state.columnsToRender
+  const columnOffsets = {}; // state.columnOffsets
+
+  // incremental way for calculating columnOffset
+  let runningOffset = columnOffsetIntervalTree.sumUntil(startIdx);
+
+  // compute column index and offsets for every columns inside the buffer
+  for (let columnIdy = startIdx; columnIdy < endIdx; columnIdy++) {
+
+    // Update the offset for rendering the column
+    columnOffsets[columnIdy] = runningOffset;
+    runningOffset += columnOffsetIntervalTree.get(columnIdy);
+
+    // Get position for the viewport column
+    const columnPosition = addColumnToBuffer(columnIdy, columnBufferSet, startIdx, endIdx, renderedColumnsCount);
+    columns[columnPosition] = columnIdy;
+  }
+
+  // now we modify the state with the newly calculated columns and offsets
+  state.columnsToRender = columns;
+  state.columnOffsets = columnOffsets;
+}
+
+/**
+ * Add the column to the buffer set if it doesn't exist.
+ * If addition isn't possible due to max buffer size, it'll replace an existing element outside the given range.
+ *
+ * @param {!number} columnIdy
+ * @param {!number} columnBufferSet
+ * @param {!number} startRange
+ * @param {!number} endRange
+ * @param {!number} maxBufferSize
+ *
+ * @return {?number} the position of the column after being added to the buffer set
+ * @private
+ */
+function addColumnToBuffer(columnIdy, columnBufferSet, startRange, endRange, maxBufferSize) {
```

Comment:
duplicated again... We can probably move this and the one in computeRenderedRows over to a helper function